Add database access resources for load testing

gabrielcorado opened the pull request:
Adds base infrastructure for load testing PostgreSQL and MySQL database access. Currently, this will only support EKS clusters.

This PR includes:
- `assets/loadtest/databases/rds/`: A Terraform definition for creating PostgreSQL and MySQL instances and IAM permissions for database agents. It will also set up a user ready for Teleport access (`alice` by default).
- `assets/loadtest/helm/database-agent/`: A Helm Chart for deploying Teleport database agents. It includes setting up the ServiceAccount with the IAM role and monitoring resources for scraping the agent metrics (which will be available for making dashboards on Grafana).

**Note: `tsh bench` charts/commands will be a separate PR (this one is already too large).**

<details>
<summary>How to use it</summary>

1. Deploy the Teleport cluster as usual. Create the EKS cluster, and Teleport auth/proxy instances.

2. Grab the EKS cluster name (same set on `cluster.yaml` file). On the `assets/loadtest/databases/rds/` run `make apply PREFX="yourname" EKS_CLUSTER_NAME="yourclustername"`. This will generate two RDS instances (PostgreSQL and MySQL) and the IAM role used by database agents. You should see something like this as a result:
    ```
    ... Apply output...
    Apply complete! Resources: 15 added, 0 changed, 0 destroyed.

    Outputs:

    database_agent_role_arn = "arn:aws:iam::000000000000:role/yourname-database-access"
    mysql_address = "yourname-00000000000000000000000000.aaaaaaaaaaaa.us-east-1.rds.amazonaws.com:3306"
    postgres_address = "yourname-00000000000000000000000000.aaaaaaaaaaaa.us-east-1.rds.amazonaws.com:5432"
    ```
3. Create a join token. On the `assets/loadtest/` and run `make create-token`.

3. Deploy the database agents. Go to `assets/loadtest/` directory and execute the `deploy-database-agents` target. Here is an example of what is necessary:

    ```
    make deploy-database-agents \
      NODE_TOKEN=<token> \
      PROXY_SERVER=yourcluster.teleportdemo.net:443 \
      NODE_REPLICAS=3 \
      DATABASE_ROLE_ARN=<database_agent_role_arn output value> \
      POSTGRES_URL=<postgres_address output value> \
      MYSQL_URL=<mysql_address output value>
    ```

</details>

gabrielcorado committed:
feat(loadtest): add database access with RDS instances

rosstimothy reviewed (commented):
The database agent helm chart looks to have a lot of duplication from the node agent helm chart. Could we consolidate to a single chart that optionally enabled ssh or db services?

smallinsky commented:
> The database agent helm chart looks to have a lot of duplication from the node agent helm chart. Could we consolidate to a single chart that optionally enabled ssh or db services?

Yea, if possible we should consolidate this but I'm thinking why we can't just use  our [teleport-kube-agent]( https://github.com/gravitational/teleport/tree/gabrielcorado/loadtest-databases/examples/chart/teleport-kube-agent) example Helm chart for DB server deployment  ?

smallinsky commented on assets/loadtest/Makefile:
```diff
@@ -76,3 +76,32 @@ forward-grafana:
+MYSQL_NAME ?= rdsmysql
+MYSQL_URL ?=
+DATABASE_ROLE_ARN ?=
+TELEPORT_VERSION ?= 13
```
Should we make this var more explicit like required as user input to prevent situation where the default 13 image version will be used ?

smallinsky commented on assets/loadtest/helm/database-agent/values.yaml:
```diff
@@ -0,0 +1,29 @@
+
+joinParams:
+  method: token
+  token_name: qwertyuiop
```
the camel_case name is inconsistent. Also by default can we just point to token file even if we will not have mounted secret ?
```suggestion
  authToken:  /etc/teleport-secrets/auth-token
```

smallinsky commented on assets/loadtest/helm/database-agent/templates/deployment.yaml:
```diff
@@ -0,0 +1,41 @@
+        app.kubernetes.io/name: {{ .Release.Name }}
+    spec:
+      serviceAccountName: {{ .Release.Name }}
+      containers:
```
probably we should have control on Memory/CPU container POD k8 resource. 

I think that if we should consider replacing this database-agent deployment with  to our `example/helm/teleport-kube-agent` chart.

smallinsky reviewed (commented):
I haven't tested it but the IaC scripts looks good to me though I have left one suggestion about using teleprot-kube-agent helm chart instead creating a new one. 

Also can we extend observability and add Grafana charts that will cover database metrics from teleport proxy/ db agent ? 

You can reach to @Tener for some chart examples.

gabrielcorado committed:
refactor(loadtest): use default teleport chart for db agents

gabrielcorado committed:
refactor(loadtest): make teleport version required for db agents

gabrielcorado commented:
@rosstimothy @smallinsky I've changed it to use the `teleport/teleport-kube-agent` to deploy database agents. The only missing part was the `PodMonitor` (for Prometheus metrics), for which I've added a script that creates it.

> Also can we extend observability and add Grafana charts that will cover database metrics from teleport proxy/ db agent ?

I planned to do it in a separate PR, but if you feel we should do it here, I can update it.

gabrielcorado committed:
fix(loadtest): set replicas count

gabrielcorado commented on assets/loadtest/databases/rds/Makefile:
```diff
@@ -0,0 +1,17 @@
+PREFIX := loadtest
+LOCAL_IP := $(shell curl -s https://checkip.amazonaws.com/)
+EKS_CLUSTER_NAME := ""
+TF_VARS = -var "region=$(REGION)" \
+		-var "prefix=$(PREFIX)" \
+		-var "local_ip=$(LOCAL_IP)" \
+		-var "eks_cluster_name=$(EKS_CLUSTER_NAME)"
```
The EKS cluster is created in a different step and does not share any value with the databases Terraform. So, users must provide the name generated (we cannot assume any default value). The other values have default values set on variables, but they're here to make it easier to customize.

gabrielcorado committed:
fix(loadtest): make databases have public access

gabrielcorado committed:
refactor(loadtest): improve deploy database agent variables check
